### PR TITLE
Prepare index before using point locator

### PR DESCRIPTION
**Pull request by kemen209**
## Description

Fix issue of not preparing index for QgsPointLocator if snapping throught QgsSnappingUtils::snapToCurrentLayer.

I use QGIS core as a library, and only use snapToCurrentLayer() and not using any snapToMap(). In this way, QgsSnappingUtils will always use its temporary locator, since the 'normal' one(can be initilzed by prepareIndex) will never been initilized.

snapToMap will not have such issue since the prepareIndex had been called properly.

### CODE REVIEW

**Comment by kemen209** [comment]
One more commit to set default value of relaxed to false(keep the same with _updateBestMatch).

**Comment by kemen209** [comment]
Another commit to fix issue on test case: 

After i adding 'prepareIndex' for 'snapToCurrentLayer', the QgsSnappingUtils will use the 'global' locator instead of temporary locator for snapping. 
```
src/core/qgssnappingutils.cpp -> QgsSnappingUtils::locatorForLayerUsingStrategy

  QgsPointLocator *loc = locatorForLayer( vl );

  if ( loc->isIndexing() || isIndexPrepared( loc, aoi ) )
    return loc;
  else
    return temporaryLocatorForLayer( vl, pointMap, tolerance );
```

And global locator use the pointer of its corresponding layer as key for its storage in QgsSnappingUtils which will survice for the whole life cycle of test case(testqgsmaptoolreverseline.cpp). 

During my debugging process, i found out that same memory address may assign to different memory-layer object of different test function(testReverseCurve, testReverseLineString, testReverseMultiLineString). Which lead to failures on the test case. 

This same-address issue only exist for this perticular test case, others, like snapToMap test case, not removing the layer at all, like qgis application, handle the layer adding and removing properly.

**Comment by kemen209** [comment]
Test Environment:
```
macOS 11.6
QGIS: master branch, 09cfa8ae0a53afe9bf01383a340b13e5c94382ce
```

I use a bash command to run 'test_app_maptoolreverseline' test case 1000 times, 5 all passed, 995 failed(same vector layer address).

```
for i in `seq 1000`; do echo "$i"; ./output/bin/test_app_maptoolreverseline; done > out1.log
```

A patch to print out the address of vector layer:
```
--- a/tests/src/app/testqgsmaptoolreverseline.cpp
+++ b/tests/src/app/testqgsmaptoolreverseline.cpp
@@ -71,6 +71,7 @@ void TestQgsMapToolReverseLine::testReverseCurve()
 {
   //create a temporary layer
   std::unique_ptr< QgsVectorLayer > memoryLayer( new QgsVectorLayer( QStringLiteral( "LineString?crs=EPSG:3946&field=pk:int" ), QStringLiteral( "vl" ), QStringLiteral( "memory" ) ) );
+  QgsDebugMsg(QString("0000000000001.1 %1").arg((long long)memoryLayer.get()));
   QVERIFY( memoryLayer->isValid() );
   QgsFeature curve( memoryLayer->dataProvider()->fields(), 1 );

@@ -107,6 +108,7 @@ void TestQgsMapToolReverseLine::testReverseLineString()
 {
   //create a temporary layer
   std::unique_ptr< QgsVectorLayer > memoryLayer( new QgsVectorLayer( QStringLiteral( "LineStringZ?crs=EPSG:3946&field=pk:int" ), QStringLiteral( "vl" ), QStringLiteral( "memory" ) ) );
+  QgsDebugMsg(QString("0000000000001.1 %1").arg((long long)memoryLayer.get()));
   QVERIFY( memoryLayer->isValid() );
   QgsFeature line( memoryLayer->dataProvider()->fields(), 1 );

@@ -143,6 +145,7 @@ void TestQgsMapToolReverseLine::testReverseMultiLineString()
 {
   //create a temporary layer
   std::unique_ptr< QgsVectorLayer > memoryLayer( new QgsVectorLayer( QStringLiteral( "MultiLineStringZ?crs=EPSG:3946&field=pk:int" ), QStringLiteral( "vl" ), QStringLiteral( "memory" ) ) );
+  QgsDebugMsg(QString("0000000000001.1 %1").arg((long long)memoryLayer.get()));
   QVERIFY( memoryLayer->isValid() );
   QgsFeature multi( memoryLayer->dataProvider()->fields(), 1 );
```

Log File:

**Comment by kemen209** [comment]
Hi, how can i retrigger thes test without a commit ?  I don't think these failure test are caused by my change. Same failure found in other's change: https://github.com/qgis/QGIS/runs/3850201375?check_suite_focus=true

**Comment by lbartoletti** [comment]
> Hi, how can i retrigger thes test without a commit ? I don't think these failure test are caused by my change. Same failure found in other's change: https://github.com/qgis/QGIS/runs/3850201375?check_suite_focus=true

Yes, some tests are flaky. It was hana first, and now 3d, but not related to your PR.

**Comment by troopa81** [comment]
> I use QGIS core as a library, and only use snapToCurrentLayer() and not using any snapToMap(). In this way, QgsSnappingUtils will always use its temporary locator, since the 'normal' one(can be initilzed by prepareIndex) will never been initilized.

Why is it an issue to use the temporaryLocator ? It will be initialized later in the *prepare* method called in all methods that need the index (nearestVertex, nearestCentroid...)

To be honest, I don't remember (It's been a while since I modified this code) why: 
- There is a need for temporary locators
- We prepare index whereas it's done in PointLocator when calling the methods using the index (nearestVertex, nearestCentroid...)

**Comment by kemen209** [comment]
> > I use QGIS core as a library, and only use snapToCurrentLayer() and not using any snapToMap(). In this way, QgsSnappingUtils will always use its temporary locator, since the 'normal' one(can be initilzed by prepareIndex) will never been initilized.
> 
> Why is it an issue to use the temporaryLocator ? It will be initialized later in the _prepare_ method called in all methods that need the index (nearestVertex, nearestCentroid...)
> 
> To be honest, I don't remember (It's been a while since I modified this code) why:
> 
> * There is a need for temporary locators
> * We prepare index whereas it's done in PointLocator when calling the methods using the index (nearestVertex, nearestCentroid...)

I have no problem with temporary locators, as a matter of fact, to avoid performance issue i had for custom environment, i remove all reference to 'normal' locator, only used temporary locator, and it did given me a large boost. And i event created a issue for help/discussion here: https://github.com/qgis/QGIS/issues/45448. 

From what i understand, the temporary locator is only for one 'temperary' snapping. Every time user request for snapping a feature, it will create a new temporary locator, and used only for small area snapping(user clicked location with small rectangle). It has no 'cache' property, unlike the 'normal' locator, which meant for using with cached/large number features.

The index start perparing from QgsSnappingUtils::prepareIndex, QgsPointLocator::prepare are only done some checking/testing.

Why am i propose this merge request? Only want to prepare index for QgsSnappingUtils::snapToCurrentLayer as did in QgsSnappingUtils::snapToMap(keep thing unified), and also because i only used snapToCurrentLayer in my scenario.

**Comment by kemen209** [comment]
> Tl;DR; As I understand, even it's not written in documentation snapToCurrentLayer is to snap quickly on one layer and wait for the result synchronously. That's why we don't prepare index.
> 
> If you want to snap on a specific layer with index prepared, you would prefer snapToMap and its options.

Sure, this is a good point on not do the perparing. Thanks. Closing it.
